### ISSUE TITLE
Memory leak in salt-master

### ISSUE THREAD

**Issue by geraint-jones-nz** [saltstack/salt#31979]
### Description of Issue/Question

I have noticed a substianaital memory leak in salt-master 

```
root@salt:~# ps axuf
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root       896  0.0  1.6 170836 33792 ?        Ss   Mar17   0:00 /usr/bin/python /usr/bin/salt-master
root      1240  0.0  1.4 246328 29472 ?        Sl   Mar17   0:00  \_ /usr/bin/python /usr/bin/salt-master
root      1241  0.0  1.5 247080 31888 ?        Sl   Mar17   0:00  \_ /usr/bin/python /usr/bin/salt-master
root      1252  0.1  3.0 451776 61828 ?        Sl   Mar17   1:50  \_ /usr/bin/python /usr/bin/salt-master
root      1321  0.1  1.6 794880 34604 ?        Sl   Mar17   1:52  \_ /usr/bin/python /usr/bin/salt-master
root      1322  0.0  1.4 170844 30336 ?        S    Mar17   0:00  \_ /usr/bin/python /usr/bin/salt-master
root      1327  0.0  1.5 547700 32176 ?        Sl   Mar17   0:00      \_ /usr/bin/python /usr/bin/salt-master
root      1330  0.0 10.3 977768 211292 ?       Sl   Mar17   0:01      \_ /usr/bin/python /usr/bin/salt-master
root      1338  0.0 10.3 999916 211612 ?       Sl   Mar17   0:01      \_ /usr/bin/python /usr/bin/salt-master
root      1339  0.0 10.3 978200 211584 ?       Sl   Mar17   0:01      \_ /usr/bin/python /usr/bin/salt-master
root      1345  0.0 10.3 978056 212132 ?       Sl   Mar17   0:01      \_ /usr/bin/python /usr/bin/salt-master
root      1350  0.0 10.3 982184 211692 ?       Sl   Mar17   0:01      \_ /usr/bin/python /usr/bin/salt-master
```

It eventually OOMs and/or kills the box.

### Setup

all states and pillars are in `gitfs` roots, otherwise pretty standard.

### Steps to Reproduce Issue

Leave the master running for 24 hrs.

### Versions Report

```
root@salt:~# salt --versions
Salt Version:
           Salt: 2015.8.7

Dependency Versions:
         Jinja2: 2.7.2
       M2Crypto: 0.21.1
           Mako: Not Installed
         PyYAML: 3.10
          PyZMQ: 14.5.0
         Python: 2.7.6 (default, Jun 22 2015, 17:58:13)
           RAET: Not Installed
        Tornado: 4.3
            ZMQ: 4.0.5
           cffi: 0.8.6
       cherrypy: 4.0.0
       dateutil: Not Installed
          gitdb: 0.5.4
      gitpython: 0.3.2 RC1
          ioflo: Not Installed
        libgit2: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.3.0
   mysql-python: Not Installed
      pycparser: 2.10
       pycrypto: 2.6.1
         pygit2: Not Installed
   python-gnupg: Not Installed
          smmap: 0.8.2
        timelib: Not Installed

System Versions:
           dist: Ubuntu 14.04 trusty
        machine: x86_64
        release: 3.13.0-79-generic
         system: Ubuntu 14.04 trusty
```

**Comment by ssgward** [comment]
@geraint-jones-nz - thanks. We're investigating

**Comment by cachedout** [comment]
Hi @geraint-jones-nz 

So, we have noticed any similar leaks over here but we have heard the occasional report of leaks that happen via the use of GitFS, so this is the first place I'm inclined to look here.

We have seen issues with GitPython in the past: https://github.com/gitpython-developers/GitPython/issues/289

I think the first thing I'd consider here is switching to the pygit2 backend, which should be a little faster anyway. Is that something you'd be willing to try? We'd also like to know if removing GitFS from the picture entirely corrects the leak, but I assume this is a machine that's in production and that's obviously easier said than done. :]

Let me know your thoughts. Thanks.

**Comment by moises-silva** [comment]
We have a similar issue, the salt-master suddenly consuming 5GB of memory for a mostly idle system with only a handful of minions. We're not using any of the git modules. We don't have pillars nor states, it's basically a default installation on C7 that we use to manually execute commands. One distinctive thing is that there are often new minions that keep trying to authenticate and never get their key accepted (those are test systems), but we're talking about ~10 minions max, so is fairly low load.

`# salt --versions
Salt Version:
           Salt: 2015.8.7
 
Dependency Versions:
         Jinja2: 2.7.2
       M2Crypto: Not Installed
           Mako: Not Installed
         PyYAML: 3.10
          PyZMQ: 14.7.0
         Python: 2.7.5 (default, Nov 20 2015, 02:00:19)
           RAET: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5
           cffi: 1.4.2
       cherrypy: Not Installed
       dateutil: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
        libgit2: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: 2.14
       pycrypto: 2.6.1
         pygit2: Not Installed
   python-gnupg: Not Installed
          smmap: Not Installed
        timelib: Not Installed
 
System Versions:
           dist: centos 7.2.1511 Core
        machine: x86_64
        release: 3.10.0-229.20.1.el7.x86_64
         system: CentOS Linux 7.2.1511 Core
`

I am considering trying to plug Heapy (http://guppy-pe.sourceforge.net/) to our salt master to be able to dump stats on memory on demand. Is there anything already in the source for memory profiling?

**Comment by ssgward** [comment]
@geraint-jones-nz / @moises-silva - do you have a minion also installed on your master or is your master just a master--no minion?

**Comment by moises-silva** [comment]
@ssgward No minion installed in my case. The master only has salt-master process started. I just checked the server again and memory is creeping up, already almost at 5GB:

```
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
15122 omega     20   0 8741292 4.899g    664 S   0.0 62.8   2:16.47 salt-master
```

I'll have to restart it soon.

**Comment by moises-silva** [comment]
In fact, this particular server has no minions at all yet. Can that become a problems somehow? it'd be kinda ironic that the heavy resource use is due to the master having no minions ...

```
$ salt-key -L
Accepted Keys:
Denied Keys:
Unaccepted Keys:
Rejected Keys:
```

**Comment by Ch3LL** [comment]
I have been trying to re-create this issue with no luck. My test case include a ubuntu14 master with 6 minions attached, including a salt-minon on the master: 

```
 7727 root      20   0  873768  48992   8072 S  0.0  1.2   0:50.73 salt-master                                                                                          
 7722 root      20   0  873496  48820   8008 S  0.0  1.2   0:50.51 salt-master                                                                                          
 7723 root      20   0  873408  48820   8008 S  0.0  1.2   0:51.06 salt-master                                                                                          
 7726 root      20   0  873412  48804   8008 S  0.0  1.2   0:50.29 salt-master                                                                                          
 7732 root      20   0  873412  48792   8008 S  0.0  1.2   0:50.93 salt-master                                                                                          
 7705 root      20   0  258628  45480   6372 S  0.0  1.1   4:27.41 salt-master                                                                                         
 7689 root      20   0  552464  40628   7916 S  0.0  1.0   1:32.33 salt-minion                                                                                          
 7698 root      20   0  242436  34892   5892 S  0.0  0.9   0:09.24 salt-master                                                                                          
 7668 root      20   0  159428  34772   8944 S  0.0  0.9   0:00.33 salt-master                                                                                          
 7712 root      20   0  536292  32032   6004 S  0.0  0.8   1:41.59 salt-master                                                                                          
 7695 root      20   0  241224  31560   5956 S  0.0  0.8   0:20.39 salt-master                                                                                          
 7708 root      20   0  159436  29504   3664 S  0.0  0.7   0:00.01 salt-master                                                                                          
 7679 root      20   0  137708  22884   8460 S  0.0  0.6   0:00.13 salt-minion                 
```

I will change my test-case to the following, based off of this new information:

1. 1 master (centos7) without any minions attached (wait a couple hours)
2. attempt to have 5 minions attach to the master but do not accept the keys
3. stop the minions after a couple hours so they are not trying to authenticate
3. wait 24 hours

@moises-silva : If the test-case above is different from how you are trying to replicate this issue please inform me.

**Comment by moises-silva** [comment]
@Ch3LL That's almost correct. But I just realized there isn't any minions even trying to connect to this system. The minions we have are still pointing to another system. That is the odd thing, the system with minions do not exhibit the problem and the system without minions is having the problem.

Having said that. I just discovered something that might be causing things to misbehave in the system with no minions (the system with high memory usage). The /etc/salt/master file had root permissions but the /etc/salt/master.d/custom file we use had set the system running user to be 'omega' (non-root). This means the master was unable to read the master configuration, it only read /etc/salt/master.d/custom. It was able to start however (I assume using built-in defaults).

The bottom line is that the system having the memory problem had wrong configuration permissions (and potentially wrong /var/cache/salt permissions), it was starting nonetheless, nothing logged (but, I guess that could be due to the wrong permissions too). Now that the permissions are corrected I'll keep an eye on the memory and see if perhaps the high memory usage was an odd side-effect of the wrong permissions, which would still be a bug, but not as bad and perhaps unrelated to what the OP reported in this issue.

I'll keep you posted.

**Comment by Ch3LL** [comment]
@moises-silva thanks for the additional information. I look forward to your results. To make sure I understand you correctly you are attempting to run salt as the user omega but you only changed the owner permissions to omega of one file:  `/etc/salt/master.d/custom` . In the meantime I am going to try to add the permissions issue to my test case and see what I see as well. 

Here is my test case and once again please correct me if i'm wrong:

1. 1 master (no minions)
2. add user omega

```
root@ch3ll-ub14:~# cat /etc/passwd | grep omega
omega:x:1001:1001::/home/omega:
```

3. add the file: /etc/salt/master.d/custom.conf and add the following settings:

```
root@ch3ll-ub14:/etc/salt/master.d# cat /etc/salt/master.d/custom.conf 
user: omega
```

4. Start master and check memory usage.

**Comment by moises-silva** [comment]
@Ch3LL That's exactly right. The ansible automation that setup this server was outdated and only created the custom.conf file with permissions for the new omega user, but didn't change the existing permissions of the other files (the service used to run as root).

**Comment by moises-silva** [comment]
Attaching a picture of the memory usage trend from last week before I restarted today.

![salt-master-memory-issue](https://cloud.githubusercontent.com/assets/680665/14055167/f0709f1a-f2b5-11e5-8954-07b2b898eb02.png)

The blue line is the free memory. The spikes on free memory are after every restart (systemctl restart salt-master). Worth noting that before I changed permissions, systemctl was unable to stop gracefully all salt-master processes and it'd hang until I did `killall -15 salt-master`. Once I fixed the permissions, the service can be stopped gracefully.

The last time the stop operation hanged, I attached gdb to one of the python processes and seemed stuck in a poll() system call inside the zmq library. Not sure if that is relevant at all.

**Comment by Ch3LL** [comment]
Thanks for verifying my test case. Just as an update, with the current test case that I have I am seeing the memory climbing now:

```
12760 omega     20   0  765564 597208   5264 S  0.0 14.8   0:04.26 salt-master                                                                                          
18821 omega     20   0  334588  37820   6024 S  7.3  0.9   0:00.22 salt-master                                                                                         
12751 omega     20   0  159748  34920   8880 S  0.0  0.9   0:01.44 salt-master                                                                                          
17312 omega     20   0  248840  34660   6012 S  0.0  0.9   0:00.20 salt-master                                                                                          
12766 omega     20   0  164180  34620   4100 S  1.0  0.9   0:44.13 salt-master                                                                                         
12767 omega     20   0  536224  31192   5288 S  0.0  0.8   0:00.61 salt-master                                                                                          
12759 omega     20   0  241136  30864   5264 S  0.0  0.8   0:00.00 salt-master                                                                                          
```

The `14.8`  is the memory%. The reason for this is the master is continually restarting due to the incorrect permissions. In the debug log I am seeing this about 12-13 times a minute:

```
2016-03-25 18:09:00,721 [salt.utils.process][INFO    ][12766] Process <class 'salt.master.MWorker'> (14877) died with exit status None, restarting...
```

@moises-silva were you seeing the above error in your log when the incorrect permissions were set? Also, thanks for including the graph before changing the permissions. Has it improved since changing the permissions correctly?

Also, I did not have any issues stopping the salt-master. Were you running as a non-root user by chance? or any other variables?

**Comment by moises-silva** [comment]
@Ch3LL No I didn't see the error, but the loglevel was defaulted to warning, and I see the log message you posted is INFO.

Regardless of that, the memory usage is climbing again even after fixing the permissions.

![salt-memory](https://cloud.githubusercontent.com/assets/680665/14057818/e719cce4-f2e1-11e5-8a77-43f436d797e9.png)

Now that I enabled debug logging, I keep seeing this every minute:

```
2016-03-25 23:30:15,562 [salt.config      ][DEBUG   ][9158] Reading configuration from /etc/salt/master
2016-03-25 23:30:15,640 [salt.config      ][DEBUG   ][9158] Guessing ID. The id can be explicitly set in /etc/salt/minion
2016-03-25 23:30:15,674 [salt.config      ][INFO    ][9158] Found minion id from generate_minion_id(): srms-01
2016-03-25 23:30:15,677 [salt.config      ][DEBUG   ][9158] Reading configuration from /etc/salt/master
2016-03-25 23:30:15,750 [salt.config      ][DEBUG   ][9158] Including configuration from '/etc/salt/master.d/rms-master.conf'
2016-03-25 23:30:15,751 [salt.config      ][DEBUG   ][9158] Reading configuration from /etc/salt/master.d/rms-master.conf
2016-03-25 23:30:16,141 [salt.utils.lazy  ][DEBUG   ][9158] LazyLoaded local_cache.clean_old_jobs
2016-03-25 23:30:16,142 [salt.utils.lazy  ][DEBUG   ][9158] Could not LazyLoad config.merge
2016-03-25 23:30:16,143 [salt.fileserver  ][DEBUG   ][9158] Updating roots fileserver cache
2016-03-25 23:30:16,144 [salt.utils.verify][DEBUG   ][9158] This salt-master instance has accepted 0 minion keys.
```

That seems odd. Is it normal the configuration file is read every minute?

**Comment by thedrow** [comment]
@moises-silva There is #23405 which discusses adding a memory profiling mode to Salt in order to debug memory leaks. You're welcome to join the discussion there.

**Comment by moises-silva** [comment]
@Ch3LL Just FYI, the process ran out of memory and was killed by the OOM:

```
[5487392.963414] [23885]   986 23885   126804     7255     124        0             0 salt-master
[5487392.964732] Out of memory: Kill process 11979 (salt-master) score 776 or sacrifice child
[5487392.965982] Killed process 11979 (salt-master) total-vm:11952560kB, anon-rss:6324848kB, file-rss:0kB
```

This is even after I fixed the permissions.

**Comment by moises-silva** [comment]
@thedrow interesting, I'll check it out and chime in if I find something else that is useful during my debugging of this memory leak.

**Comment by Ch3LL** [comment]
@moises-silva I am wondering if the master is restarting. Is that the entire debug output that occurs every minute? 

Also how did you set the permissions for the user omega to run the salt-master?

**Comment by moises-silva** [comment]
@Ch3LL I don't think the master is restarting. I've monitored the PID number of all salt-master processes and is not changing, so they are not restarting. I think this is some sort of reconfiguration feature, checking for changes in the configuration every minute?

To answer your question. I did the following to fix the permissions a few days ago when I posted here:

chown -R omega:omega /etc/salt
chown -R omega:omega /var/cache/salt

Yesterday I provisioned the salt-master code to use [Pympler tracking module](https://pythonhosted.org/Pympler/muppy.html#the-tracker-module) to list the python object count every minute as part of the configuration debug output I was seeing (I took the idea of using the Pympler module from issue #23405. However, now I'm not sure if I am losing my mind, but now it seems memory has been stable. I don't see how my code could affect the memory usage, unless the Pympler module somehow changes gc settings or something else that masks the problem?

I'll remove the Pympler code I added and see if the leak comes back.

Just for reference, these are the modifications I did to add Pympler to the debug output: [memdebug.diff](https://gist.github.com/moises-silva/65eed9f083cb48aa453d)

**Comment by Ch3LL** [comment]
@moises-silva thanks for the additional information. I did compare against a good salt-master and those logs look normal to see every minute, especially since you are not seeing any additional logs about the master restarting.

Concerning your permissions, looks like you also need to give read and write to all of the following directories:

*/etc/salt
*/var/cache/salt
*/var/log/salt
*/var/run/salt

You can run the following command: `chown -R user /etc/salt /var/cache/salt /var/log/salt /var/run/salt` Its all documented [here](https://docs.saltstack.com/en/latest/ref/configuration/nonroot.html)

Can you please make sure the user has read and write for and owns those directories above and check your results once again?

I am going to change my test scenario as well to setup the correct permissions and see if the memory increases.

**Comment by Ch3LL** [comment]
@moises-silva  after changing to the correct permissions in my test scenario I'm still not seeing the memory increasing. How about on your end? Is the memory not increasing/leak anymore still?

**Comment by moises-silva** [comment]
@Ch3LL Sorry for the delay. I have not seen the issue again. Several things have changed though. That system now has minions connected and we have a process that automatically accepts the minion keys (based on some out of band authentication mechanism). At this point I don't know if the permissions or the fact that there's minions and keys in the system made the difference, or something else. For now I can just keep an eye on the system and see if the issue presents itself again.

Thanks for all the help.

**Comment by Ch3LL** [comment]
@moises-silva thanks for the update. I'm glad everything is working for you now. But please update the ticket if that changes. @geraint-jones-nz were you able to try the recommendation from @cachedout  to switch to pygit2? Are you still seeing this issue?

**Comment by cachedout** [comment]
@moises-silva Hmm, given your latest comment this sounds like a leak I identified and fixed yesterday.

https://github.com/saltstack/salt/pull/32373

**Comment by moises-silva** [comment]
@cachedout Ah! that certainly seems possible. Thank you!
